src/vsr.zig

```diff
@@ -1917,3 +1918,52 @@ const ViewChangeHeadersArray = struct {
         headers.array.appendAssumeCapacity(Headers.dvc_blank(op));
     }
 };
+
+pub const Op = struct {
+    pub fn checkpoint_before(checkpoint: u64) ?u64 {
+        assert(checkpoint_valid(checkpoint));
+
+        if (checkpoint == 0) {
+            return null;
+        }
+
+        if (checkpoint == constants.journal_slot_count - constants.lsm_batch_multiple - 1) {
+            return 0;
+        }
+
+        return checkpoint - (constants.journal_slot_count - constants.lsm_batch_multiple);
+    }
+
+    pub fn checkpoint_after(checkpoint: u64) u64 {
+        assert(checkpoint_valid(checkpoint));
+
+        const checkpoint_next = op: {
+            if (checkpoint == 0) {
+                // First wrap: op_checkpoint_next = 8-2-1 = 5
+                break :op constants.journal_slot_count - constants.lsm_batch_multiple - 1;
+            } else {
+                // Second wrap: op_checkpoint_next = 5+8-2 = 11
+                // Third wrap: op_checkpoint_next = 11+8-2 = 17
+                break :op checkpoint + constants.journal_slot_count - constants.lsm_batch_multiple;
+            }
+        };
+
+        assert((checkpoint_next + 1) % constants.lsm_batch_multiple == 0);
+        assert(checkpoint_valid(checkpoint_next));
+        assert(checkpoint_before(checkpoint_next) == checkpoint);
+
+        return checkpoint_next;
```

Comment:
```suggestion
        return result;
```

`checkpoint_after` and `checkpoint_next` are different only because of shadowing. Therefore, use a simpler local variable like `result` or `op`.